### ISSUE TITLE
Reconcile doesn't fire sometimes after adding a finalizer

### ISSUE THREAD

**Issue by ahmelsayed** [kedacore/keda#437]
@zroubalik I'm not sure what the issue is here, and I might need your help. 

I have 2 nightly tests that deploy those yamls ([test 1](https://github.com/kedacore/keda/blob/master/tests/scalers/azure-queue.test.ts#L103-L159), [test 2](https://github.com/kedacore/keda/blob/master/tests/scalers/azure-queue-trigger-auth.test.ts#L109-L172)) 

In test 1 the ScaledObject looks like this

```yaml
apiVersion: keda.k8s.io/v1alpha1
kind: ScaledObject
metadata:
  name: test-scaledobject
  labels:
    deploymentName: test-deployment
spec:
  scaleTargetRef:
    deploymentName: test-deployment
  pollingInterval: 5
  maxReplicaCount: 4
  cooldownPeriod: 10
  triggers:
  - type: azure-queue
    metadata:
      queueName: queue-name
      connection: AzureWebJobsStorage
```

and in test 2 it looks like this

```yaml
apiVersion: keda.k8s.io/v1alpha1
kind: ScaledObject
metadata:
  name: test-scaledobject
  labels:
    deploymentName: test-deployment
spec:
  scaleTargetRef:
    deploymentName: test-deployment
  pollingInterval: 5
  maxReplicaCount: 1
  cooldownPeriod: 10
  triggers:
  - type: azure-queue
    authenticationRef:
      name: azure-queue-auth
    metadata:
      queueName: queue-name
```

In the first call for `Reconcile()`
https://github.com/kedacore/keda/blob/99b674c68c0d7b912c3fb2d99fab01b59ebebef4/pkg/controller/scaledobject/scaledobject_controller.go#L91

It matches this condition. 
https://github.com/kedacore/keda/blob/99b674c68c0d7b912c3fb2d99fab01b59ebebef4/pkg/controller/scaledobject/scaledobject_controller.go#L134-L139

`.addFinalizer` calls `Update` on the ScaledObject, then this branch returns. I'm assuming that update should trigger another `Reconcile()` call. And it does for test 1, **but not test 2**. 

I'm not sure what the difference is.

this change will fix it 

```diff
if !contains(scaledObject.GetFinalizers(), scaledObjectFinalizer) {
	if err := r.addFinalizer(reqLogger, scaledObject); err != nil {
		return reconcile.Result{}, err
	}
-	return reconcile.Result{}, nil
+	return reconcile.Result{Requeue: true}, nil
}
```

Though it'll trigger Reconcile twice for test 1. I don't see anything in the logs. Not sure if you might know what the issue is. if not I can debug through the operator-sdk

**Comment by zroubalik** [comment]
thanks for letting me know @ahmelsayed I'll take a look what's the problem

**Comment by balchua** [comment]
I am also experiencing this one, while working on the Redis scaler with TriggerAuthentication.

**Comment by zroubalik** [comment]
TLDR: optional structures in ScaledObject type should be pointers, to prevent client with initializing them with an update.


So the problem is in inconsistency `ScaledObject` type, sometimes adding a finalizer on the object incremented Object's `Generation`, sometimes not -> that' why the Reconcile loop doesn't fire sometimes.

Adding a finalizer should not increment Generation. But when we [update](https://github.com/kedacore/keda/blob/master/pkg/controller/scaledobject/scaledobject_finalizer.go#L45) the object with finalizer, sometimes in the `ScaledObject` the optional structures (eg. authenticationRef) weren't defined, so the client intialized them which caused increment in Object's `Generation`, this could be prevented if the optional structures were pointers.
